File: backend/app/engine/indicators.py
"""
技术指标计算模块
纯 Python 实现，无需外部依赖
"""

from typing import List, Optional


def calculate_sma(data: List[float], period: int) -> Optional[float]:
    """
    计算简单移动平均线 (SMA)

    Args:
        data: 价格数据列表（最新的在最后）
        period: 计算周期

    Returns:
        SMA 值，数据不足时返回 None
    """
    if len(data) < period:
        return None

    return sum(data[-period:]) / period


def calculate_ema(data: List[float], period: int) -> Optional[float]:
    """
    计算指数移动平均线 (EMA)

    Args:
        data: 价格数据列表（最新的在最后）
        period: 计算周期

    Returns:
        EMA 值，数据不足时返回 None
    """
    if len(data) < period:
        return None

    # EMA 公式: EMA = Price(t) * k + EMA(y) * (1 - k)
    # k = 2 / (N + 1)
    k = 2 / (period + 1)

    # 初始 EMA 使用 SMA
    ema = sum(data[:period]) / period

    # 计算后续 EMA
    for price in data[period:]:
        ema = price * k + ema * (1 - k)

    return ema


def calculate_rsi(data: List[float], period: int = 14) -> Optional[float]:
    """
    计算相对强弱指数 (RSI)

    Args:
        data: 价格数据列表（最新的在最后）
        period: 计算周期，默认 14

    Returns:
        RSI 值 (0-100)，数据不足时返回 None
    """
    if len(data) < period + 1:
        return None

    # 计算价格变化
    changes = [data[i] - data[i - 1] for i in range(1, len(data))]

    # 分离上涨和下跌
    gains = [max(change, 0) for change in changes]
    losses = [abs(min(change, 0)) for change in changes]

    # 计算平均上涨和平均下跌
    avg_gain = sum(gains[-period:]) / period
    avg_loss = sum(losses[-period:]) / period

    if avg_loss == 0:
        return 100.0

    rs = avg_gain / avg_loss
    rsi = 100 - (100 / (1 + rs))

    return rsi


def calculate_bollinger_bands(
    data: List[float],
    period: int = 20,
    std_dev: float = 2.0,
) -> Optional[dict]:
    """
    计算布林带 (Bollinger Bands)

    Args:
        data: 价格数据列表（最新的在最后）
        period: 计算周期，默认 20
        std_dev: 标准差倍数，默认 2.0

    Returns:
        包含 upper, middle, lower 的字典，数据不足时返回 None
    """
    if len(data) < period:
        return None

    # 中轨是 SMA
    prices = data[-period:]
    middle = sum(prices) / period

    # 计算标准差
    variance = sum((p - middle) ** 2 for p in prices) / period
    std = variance ** 0.5

    upper = middle + std_dev * std
    lower = middle - std_dev * std

    return {
        "upper": upper,
        "middle": middle,
        "lower": lower,
    }


def check_ma_cross(
    data: List[float],
    fast_period: int,
    slow_period: int,
) -> Optional[str]:
    """
    检查均线交叉

    Args:
        data: 价格数据列表（最新的在最后）
        fast_period: 快速均线周期
        slow_period: 慢速均线周期

    Returns:
        "golden" - 金叉（快速上穿慢速）
        "death" - 死叉（快速下穿慢速）
        None - 无交叉
    """
    if len(data) < slow_period + 1:
        return None

    # 计算当前和前一个的均线
    current_fast = sum(data[-fast_period:]) / fast_period
    current_slow = sum(data[-slow_period:]) / slow_period

    prev_fast = sum(data[-fast_period - 1:-1]) / fast_period
    prev_slow = sum(data[-slow_period - 1:-1]) / slow_period

    # 检查交叉
    if prev_fast <= prev_slow and current_fast > current_slow:
        return "golden"  # 金叉
    elif prev_fast >= prev_slow and current_fast < current_slow:
        return "death"  # 死叉

    return None


def check_bollinger_touch(
    data: List[float],
    period: int = 20,
    std_dev: float = 2.0,
) -> Optional[str]:
    """
    检查布林带突破

    Args:
        data: 价格数据列表（最新的在最后）
        period: 计算周期
        std_dev: 标准差倍数

    Returns:
        "above_upper" - 突破上轨
        "below_lower" - 突破下轨
        None - 未突破
    """
    if len(data) < period:
        return None

    bands = calculate_bollinger_bands(data, period, std_dev)
    if not bands:
        return None

    current_price = data[-1]

    if current_price > bands["upper"]:
        return "above_upper"
    elif current_price < bands["lower"]:
        return "below_lower"

    return None


def calculate_macd(
    data: List[float],
    fast: int = 12,
    slow: int = 26,
    signal: int = 9,
) -> Optional[dict]:
    """计算 MACD"""
    if len(data) < slow + signal:
        return None

    def ema_series(prices: List[float], period: int) -> List[float]:
        k = 2 / (period + 1)
        result = [sum(prices[:period]) / period]
        for price in prices[period:]:
            result.append(price * k + result[-1] * (1 - k))
        return result

    ema_fast = ema_series(data, fast)
    # 对齐长度：slow 的 EMA 从索引 slow-1 开始
    ema_slow = ema_series(data, slow)
    offset = slow - fast
    dif = [f - s for f, s in zip(ema_fast[offset:], ema_slow)]

    dea_k = 2 / (signal + 1)
    dea = [sum(dif[:signal]) / signal]
    for v in dif[signal:]:
        dea.append(v * dea_k + dea[-1] * (1 - dea_k))

    # 对齐到 dif
    dea_aligned = [None] * (signal - 1) + dea
    histogram = [
        (d - e) * 2 if e is not None else None
        for d, e in zip(dif, dea_aligned)
    ]

    return {
        "dif": dif[-1],
        "dea": dea[-1],
        "histogram": histogram[-1],
        "prev_dif": dif[-2] if len(dif) >= 2 else None,
        "prev_dea": dea[-2] if len(dea) >= 2 else None,
    }


def check_macd_signal(
    data: List[float],
    fast: int = 12,
    slow: int = 26,
    signal: int = 9,
) -> Optional[str]:
    """检查 MACD 信号"""
    result = calculate_macd(data, fast, slow, signal)
    if not result:
        return None

    dif, dea = result["dif"], result["dea"]
    prev_dif, prev_dea = result["prev_dif"], result["prev_dea"]
    histogram = result["histogram"]

    if prev_dif is not None and prev_dea is not None:
        if prev_dif <= prev_dea and dif > dea:
            return "golden"
        if prev_dif >= prev_dea and dif < dea:
            return "death"

    if histogram is not None:
        if histogram > 0:
            return "above_zero"
        if histogram < 0:
            return "below_zero"

    return None


def calculate_kdj(
    highs: List[float],
    lows: List[float],
    closes: List[float],
    period: int = 9,
) -> Optional[dict]:
    """计算 KDJ"""
    if len(closes) < period + 1:
        return None

    rsv_list = []
    for i in range(period - 1, len(closes)):
        h = max(highs[i - period + 1: i + 1])
        l = min(lows[i - period + 1: i + 1])
        rsv = (closes[i] - l) / (h - l) * 100 if h != l else 50
        rsv_list.append(rsv)

    k, d = 50.0, 50.0
    prev_k, prev_d = 50.0, 50.0
    for rsv in rsv_list:
        prev_k, prev_d = k, d
        k = k * 2 / 3 + rsv / 3
        d = d * 2 / 3 + k / 3

    j = 3 * k - 2 * d
    return {"k": k, "d": d, "j": j, "prev_k": prev_k, "prev_d": prev_d}


def check_kdj_signal(
    highs: List[float],
    lows: List[float],
    closes: List[float],
    period: int = 9,
) -> Optional[str]:
    """检查 KDJ 信号"""
    result = calculate_kdj(highs, lows, closes, period)
    if not result:
        return None

    k, d = result["k"], result["d"]
    prev_k, prev_d = result["prev_k"], result["prev_d"]

    if prev_k <= prev_d and k > d:
        return "k_cross_up"
    if prev_k >= prev_d and k < d:
        return "k_cross_down"
    if k > 80:
        return "overbought"
    if k < 20:
        return "oversold"

    return None


def check_volume_spike(
    volumes: List[float],
    ma_period: int = 20,
    multiplier: float = 1.5,
) -> bool:
    """检查成交量放大"""
    if len(volumes) < ma_period + 1:
        return False
    avg = sum(volumes[-ma_period - 1:-1]) / ma_period
    return volumes[-1] > avg * multiplier if avg > 0 else False


class IndicatorCalculator:
    """指标计算器"""

    def __init__(self, klines: List[dict]):
        """
        Args:
            klines: K 线数据列表，每项包含 open, high, low, close, volume
        """
        self.klines = klines
        self.closes = [k["close"] for k in klines]
        self.highs = [k["high"] for k in klines]
        self.lows = [k["low"] for k in klines]
        self.volumes = [k["volume"] for k in klines]

    def rsi(self, period: int = 14) -> Optional[float]:
        return calculate_rsi(self.closes, period)

    def sma(self, period: int) -> Optional[float]:
        return calculate_sma(self.closes, period)

    def ema(self, period: int) -> Optional[float]:
        return calculate_ema(self.closes, period)

    def bollinger(self, period: int = 20, std_dev: float = 2.0) -> Optional[dict]:
        return calculate_bollinger_bands(self.closes, period, std_dev)

    def ma_cross(self, fast: int, slow: int) -> Optional[str]:
        return check_ma_cross(self.closes, fast, slow)

    def bollinger_touch(self, period: int = 20, std_dev: float = 2.0) -> Optional[str]:
        return check_bollinger_touch(self.closes, period, std_dev)

    def macd_signal(self, fast: int = 12, slow: int = 26, signal: int = 9) -> Optional[str]:
        return check_macd_signal(self.closes, fast, slow, signal)

    def kdj_signal(self, period: int = 9) -> Optional[str]:
        return check_kdj_signal(self.highs, self.lows, self.closes, period)

    def volume_spike(self, ma_period: int = 20, multiplier: float = 1.5) -> bool:
        return check_volume_spike(self.volumes, ma_period, multiplier)

    def price_change_pct(self) -> Optional[float]:
        """最新K线涨跌幅（%）"""
        if len(self.closes) < 2:
            return None
        prev = self.closes[-2]
        return (self.closes[-1] - prev) / prev * 100 if prev else None

    def current_price(self) -> float:
        return self.closes[-1] if self.closes else 0.0

File: frontend/src/components/kline-chart/indicators.ts
/**
 * K线图技术指标配置
 */

// MA指标配置
export const maIndicator = {
  name: 'MA',
  shortName: 'MA',
  calcParams: [5, 10, 20, 60],
  figures: [
    { key: 'ma5', title: 'MA5: ', type: 'line' as const },
    { key: 'ma10', title: 'MA10: ', type: 'line' as const },
    { key: 'ma20', title: 'MA20: ', type: 'line' as const },
    { key: 'ma60', title: 'MA60: ', type: 'line' as const },
  ],
  calc: (kLineDataList: any[], indicator: any) => {
    const params = indicator.calcParams || [5, 10, 20, 60];
    return kLineDataList.map((kLineData, i) => {
      const result: Record<string, number | undefined> = {};
      params.forEach((period: number) => {
        if (i >= period - 1) {
          let sum = 0;
          for (let j = 0; j < period; j++) {
            sum += kLineDataList[i - j].close;
          }
          result[`ma${period}`] = sum / period;
        } else {
          result[`ma${period}`] = undefined;
        }
      });
      return result;
    });
  },
};

// MACD指标配置
export const macdIndicator = {
  name: 'MACD',
  shortName: 'MACD',
  calcParams: [12, 26, 9],
  figures: [
    { key: 'dif', title: 'DIF: ', type: 'line' as const },
    { key: 'dea', title: 'DEA: ', type: 'line' as const },
    { key: 'macd', title: 'MACD: ', type: 'bar' as const },
  ],
  calc: (kLineDataList: any[], indicator: any) => {
    const params = indicator.calcParams || [12, 26, 9];
    const [fast, slow, signal] = params;
    const closes = kLineDataList.map((d: any) => d.close);
    
    const emaFast: number[] = [];
    const emaSlow: number[] = [];
    const dif: (number | undefined)[] = [];
    const dea: (number | undefined)[] = [];
    const macd: (number | undefined)[] = [];
    
    closes.forEach((close: number, i: number) => {
      if (i === 0) {
        emaFast.push(close);
        emaSlow.push(close);
      } else {
        const kFast = 2 / (fast + 1);
        const kSlow = 2 / (slow + 1);
        emaFast.push(close * kFast + emaFast[i - 1] * (1 - kFast));
        emaSlow.push(close * kSlow + emaSlow[i - 1] * (1 - kSlow));
      }
      dif.push(emaFast[i] - emaSlow[i]);
    });
    
    dif.forEach((d, i) => {
      if (i === 0) {
        dea.push(d);
      } else {
        const kSignal = 2 / (signal + 1);
        dea.push(d! * kSignal + dea[i - 1]! * (1 - kSignal));
      }
      macd.push((dif[i]! - dea[i]!) * 2);
    });
    
    return kLineDataList.map((_: any, i: number) => ({
      dif: dif[i],
      dea: dea[i],
      macd: macd[i],
    }));
  },
};

// KDJ指标配置
export const kdjIndicator = {
  name: 'KDJ',
  shortName: 'KDJ',
  calcParams: [9, 3, 3],
  figures: [
    { key: 'k', title: 'K: ', type: 'line' as const },
    { key: 'd', title: 'D: ', type: 'line' as const },
    { key: 'j', title: 'J: ', type: 'line' as const },
  ],
  calc: (kLineDataList: any[], indicator: any) => {
    const params = indicator.calcParams || [9, 3, 3];
    const [n, m1, m2] = params;
    const k: (number | undefined)[] = [];
    const d: (number | undefined)[] = [];
    const j: (number | undefined)[] = [];
    
    kLineDataList.forEach((data, i) => {
      if (i < n - 1) {
        k.push(undefined);
        d.push(undefined);
        j.push(undefined);
        return;
      }
      
      let low = data.low;
      let high = data.high;
      for (let idx = 0; idx < n; idx++) {
        low = Math.min(low, kLineDataList[i - idx].low);
        high = Math.max(high, kLineDataList[i - idx].high);
      }
      const rsv = high === low ? 50 : (data.close - low) / (high - low) * 100;
      
      if (i === n - 1) {
        k.push(50);
        d.push(50);
      } else {
        const prevK = k[i - 1] ?? 50;
        const prevD = d[i - 1] ?? 50;
        k.push(prevK * (m1 - 1) / m1 + rsv / m1);
        d.push(prevD * (m2 - 1) / m2 + k[i]! / m2);
      }
      j.push(3 * k[i]! - 2 * d[i]!);
    });
    
    return kLineDataList.map((_: any, i: number) => ({
      k: k[i],
      d: d[i],
      j: j[i],
    }));
  },
};

// RSI指标配置
export const rsiIndicator = {
  name: 'RSI',
  shortName: 'RSI',
  calcParams: [14],
  figures: [
    { key: 'rsi', title: 'RSI(14): ', type: 'line' as const },
  ],
  calc: (kLineDataList: any[], indicator: any) => {
    const params = indicator.calcParams || [14];
    const [period] = params;
    const rsi: (number | undefined)[] = [];
    let avgGain = 0;
    let avgLoss = 0;
    
    kLineDataList.forEach((data, i) => {
      if (i === 0) {
        rsi.push(undefined);
        return;
      }
      
      const change = data.close - kLineDataList[i - 1].close;
      const gain = change > 0 ? change : 0;
      const loss = change < 0 ? -change : 0;
      
      if (i < period) {
        avgGain += gain / period;
        avgLoss += loss / period;
        rsi.push(undefined);
      } else if (i === period) {
        avgGain = (avgGain * (period - 1) + gain) / period;
        avgLoss = (avgLoss * (period - 1) + loss) / period;
        const rs = avgLoss === 0 ? 100 : avgGain / avgLoss;
        rsi.push(100 - 100 / (1 + rs));
      } else {
        avgGain = (avgGain * (period - 1) + gain) / period;
        avgLoss = (avgLoss * (period - 1) + loss) / period;
        const rs = avgLoss === 0 ? 100 : avgGain / avgLoss;
        rsi.push(100 - 100 / (1 + rs));
      }
    });
    
    return kLineDataList.map((_: any, i: number) => ({
      rsi: rsi[i],
    }));
  },
};

// BOLL指标配置
export const bollIndicator = {
  name: 'BOLL',
  shortName: 'BOLL',
  calcParams: [20, 2],
  figures: [
    { key: 'upper', title: 'UP: ', type: 'line' as const },
    { key: 'middle', title: 'MID: ', type: 'line' as const },
    { key: 'lower', title: 'LOW: ', type: 'line' as const },
  ],
  calc: (kLineDataList: any[], indicator: any) => {
    const params = indicator.calcParams || [20, 2];
    const [period, stdDev] = params;
    
    return kLineDataList.map((data, i) => {
      if (i < period - 1) {
        return { upper: undefined, middle: undefined, lower: undefined };
      }
      
      let sum = 0;
      for (let j = 0; j < period; j++) {
        sum += kLineDataList[i - j].close;
      }
      const middle = sum / period;
      
      let variance = 0;
      for (let j = 0; j < period; j++) {
        variance += Math.pow(kLineDataList[i - j].close - middle, 2);
      }
      const std = Math.sqrt(variance / period);
      
      return {
        upper: middle + stdDev * std,
        middle,
        lower: middle - stdDev * std,
      };
    });
  },
};

// 成交量指标
export const volumeIndicator = {
  name: 'VOL',
  shortName: 'VOL',
  figures: [
    { key: 'volume', title: 'V: ', type: 'bar' as const },
  ],
  calc: (kLineDataList: any[]) => {
    return kLineDataList.map((data: any) => ({
      volume: data.volume,
    }));
  },
};

// 获取启用的指标列表
export function getEnabledIndicators(config: {
  ma?: boolean | number[];
  macd?: boolean;
  kdj?: boolean;
  rsi?: boolean;
  boll?: boolean;
  volume?: boolean;
}) {
  const indicators: any[] = [];
  
  if (config.ma) {
    const periods = Array.isArray(config.ma) ? config.ma : [5, 10, 20, 60];
    indicators.push({
      ...maIndicator,
      calcParams: periods,
    });
  }
  
  if (config.boll !== false) {
    indicators.push(bollIndicator);
  }
  
  if (config.volume !== false) {
    indicators.push(volumeIndicator);
  }
  
  if (config.macd) {
    indicators.push(macdIndicator);
  }
  
  if (config.kdj) {
    indicators.push(kdjIndicator);
  }
  
  if (config.rsi) {
    indicators.push(rsiIndicator);
  }
  
  return indicators;
}
